File: Android_GUI/src/com/github/chenxiaolong/dualbootpatcher/switcher/RomNameInputDialog.java
/*
 * Copyright (C) 2014  Andrew Gunnerson <andrewgunnerson@gmail.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package com.github.chenxiaolong.dualbootpatcher.switcher;

import android.app.Dialog;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v4.app.DialogFragment;
import android.support.v4.app.Fragment;
import android.view.Window;
import android.view.WindowManager;
import android.widget.EditText;
import android.widget.TextView;

import com.afollestad.materialdialogs.DialogAction;
import com.afollestad.materialdialogs.MaterialDialog;
import com.afollestad.materialdialogs.MaterialDialog.SingleButtonCallback;
import com.github.chenxiaolong.dualbootpatcher.R;
import com.github.chenxiaolong.dualbootpatcher.RomUtils.RomInformation;

public class RomNameInputDialog extends DialogFragment {
    private static final String ARG_ROM = "rom";

    public interface RomNameInputDialogListener {
        void onRomNameChanged(String newName);
    }

    public static RomNameInputDialog newInstanceFromFragment(Fragment parent, RomInformation info) {
        if (parent != null) {
            if (!(parent instanceof RomNameInputDialogListener)) {
                throw new IllegalStateException(
                        "Parent fragment must implement RomNameInputDialogListener");
            }
        }

        RomNameInputDialog frag = new RomNameInputDialog();
        frag.setTargetFragment(parent, 0);
        Bundle args = new Bundle();
        args.putParcelable(ARG_ROM, info);
        frag.setArguments(args);
        return frag;
    }

    public static RomNameInputDialog newInstanceFromActivity(RomInformation info) {
        RomNameInputDialog frag = new RomNameInputDialog();
        Bundle args = new Bundle();
        args.putParcelable(ARG_ROM, info);
        frag.setArguments(args);
        return frag;
    }

    RomNameInputDialogListener getOwner() {
        Fragment f = getTargetFragment();
        if (f == null) {
            if (!(getActivity() instanceof RomNameInputDialogListener)) {
                throw new IllegalStateException(
                        "Activity must implement RomNameInputDialogListener");
            }
            return (RomNameInputDialogListener) getActivity();
        } else {
            return (RomNameInputDialogListener) getTargetFragment();
        }
    }

    @Override
    @NonNull
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        RomInformation info = getArguments().getParcelable(ARG_ROM);

        String title = String.format(getString(R.string.rename_rom_title),
                info.getDefaultName());
        String message = String.format(getString(R.string.rename_rom_desc),
                info.getDefaultName());

        Dialog dialog = new MaterialDialog.Builder(getActivity())
                .title(title)
                .customView(R.layout.dialog_textbox, true)
                .positiveText(R.string.ok)
                .negativeText(R.string.cancel)
                .onPositive(new SingleButtonCallback() {
                    @Override
                    public void onClick(@NonNull MaterialDialog dialog,
                                        @NonNull DialogAction which) {
                        RomNameInputDialogListener owner = getOwner();
                        if (owner == null) {
                            return;
                        }

                        EditText et = (EditText) dialog.findViewById(R.id.edittext);
                        String newName = et.getText().toString().trim();

                        if (newName.isEmpty()) {
                            owner.onRomNameChanged(null);
                        } else {
                            owner.onRomNameChanged(newName);
                        }
                    }
                })
                .build();

        TextView tv = ((MaterialDialog) dialog).getCustomView().findViewById(R.id.message);
        tv.setText(message);

        setCancelable(false);
        dialog.setCanceledOnTouchOutside(false);

        // Show keyboard
        Window window = dialog.getWindow();
        window.setSoftInputMode(WindowManager.LayoutParams.SOFT_INPUT_STATE_ALWAYS_VISIBLE);

        return dialog;
    }
}

File: Android_GUI/src/main/java/com/github/chenxiaolong/dualbootpatcher/switcher/RomNameInputDialog.kt
/*
 * Copyright (C) 2014-2019  Andrew Gunnerson <andrewgunnerson@gmail.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package com.github.chenxiaolong.dualbootpatcher.switcher

import android.app.Dialog
import android.os.Bundle
import android.view.WindowManager
import androidx.fragment.app.DialogFragment
import androidx.fragment.app.Fragment
import com.afollestad.materialdialogs.MaterialDialog
import com.afollestad.materialdialogs.input.input
import com.github.chenxiaolong.dualbootpatcher.R
import com.github.chenxiaolong.dualbootpatcher.RomUtils.RomInformation

class RomNameInputDialog : DialogFragment() {
    internal val owner: RomNameInputDialogListener?
        get() {
            val f = targetFragment
            return if (f == null) {
                if (activity !is RomNameInputDialogListener) {
                    throw IllegalStateException(
                            "Activity must implement RomNameInputDialogListener")
                }
                activity as RomNameInputDialogListener?
            } else {
                targetFragment as RomNameInputDialogListener?
            }
        }

    interface RomNameInputDialogListener {
        fun onRomNameChanged(newName: String?)
    }

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        val info = arguments!!.getParcelable<RomInformation>(ARG_ROM)!!

        val title = String.format(getString(R.string.rename_rom_title), info.defaultName)
        val message = String.format(getString(R.string.rename_rom_desc), info.defaultName)

        val dialog = MaterialDialog(requireActivity())
                .title(text = title)
                .message(text = message)
                .positiveButton(R.string.ok)
                .negativeButton(R.string.cancel)
                .input(allowEmpty = true) { _, text ->
                    val owner = owner ?: return@input
                    val newName = text.toString().trim { it <= ' ' }

                    if (newName.isEmpty()) {
                        owner.onRomNameChanged(null)
                    } else {
                        owner.onRomNameChanged(newName)
                    }
                }

        isCancelable = false
        dialog.setCanceledOnTouchOutside(false)

        // Show keyboard
        val window = dialog.window
        window!!.setSoftInputMode(WindowManager.LayoutParams.SOFT_INPUT_STATE_ALWAYS_VISIBLE)

        return dialog
    }

    companion object {
        private const val ARG_ROM = "rom"

        fun newInstanceFromFragment(parent: Fragment?, info: RomInformation): RomNameInputDialog {
            if (parent != null) {
                if (parent !is RomNameInputDialogListener) {
                    throw IllegalStateException(
                            "Parent fragment must implement RomNameInputDialogListener")
                }
            }

            val frag = RomNameInputDialog()
            frag.setTargetFragment(parent, 0)
            val args = Bundle()
            args.putParcelable(ARG_ROM, info)
            frag.arguments = args
            return frag
        }

        fun newInstanceFromActivity(info: RomInformation): RomNameInputDialog {
            val frag = RomNameInputDialog()
            val args = Bundle()
            args.putParcelable(ARG_ROM, info)
            frag.arguments = args
            return frag
        }
    }
}
